lib/propolis/src/cpuid.rs

```diff
@@ -0,0 +1,413 @@
+        self.map.insert(ident, entry)
+    }
+    pub fn remove(&mut self, ident: Ident) -> Option<Entry> {
+        self.map.remove(&ident)
+    }
+    pub fn remove_all(&mut self, func: u32) {
+        self.map.retain(|ident, _val| ident.0 != func);
+    }
+    pub fn get(&self, ident: Ident) -> Option<&Entry> {
+        self.map.get(&ident)
+    }
+    pub fn get_mut(&mut self, ident: Ident) -> Option<&mut Entry> {
+        self.map.get_mut(&ident)
+    }
+    pub fn is_empty(&self) -> bool {
+        self.map.is_empty()
+    }
+
+    pub fn iter(&self) -> Iter {
+        Iter(self.map.iter())
+    }
+
+    pub fn for_regs(&self, eax: u32, ecx: u32) -> Option<Entry> {
+        if let Some(ent) = self.map.get(&Ident(eax, Some(ecx))) {
+            // Exact match
+            Some(*ent)
+        } else if let Some(ent) = self.map.get(&Ident(eax, None)) {
+            // Function-only match
+            Some(*ent)
+        } else {
+            None
+        }
+    }
+}
+impl From<Set> for Vec<bhyve_api::vcpu_cpuid_entry> {
+    fn from(value: Set) -> Self {
+        let mut out = Vec::with_capacity(value.map.len());
+        out.extend(value.map.iter().map(|(ident, leaf)| {
+            let vce_flags = match ident.1.as_ref() {
+                Some(_) => bhyve_api::VCE_FLAG_MATCH_INDEX,
+                None => 0,
+            };
+            bhyve_api::vcpu_cpuid_entry {
+                vce_function: ident.0,
+                vce_index: ident.1.unwrap_or(0),
+                vce_flags,
+                vce_eax: leaf.eax,
+                vce_ebx: leaf.ebx,
+                vce_ecx: leaf.ecx,
+                vce_edx: leaf.edx,
+                ..Default::default()
+            }
+        }));
+        out
+    }
+}
+
+/// Convert a [vcpu_cpuid_entry](bhyve_api::vcpu_cpuid_entry) into an ([Ident],
+/// [Entry]) tuple, suitable for insertion into a [Set].
+///
+/// This would be implemented as a [From] trait if rust let us.
+pub fn from_raw(value: bhyve_api::vcpu_cpuid_entry) -> (Ident, Entry) {
+    let idx = if value.vce_flags & bhyve_api::VCE_FLAG_MATCH_INDEX != 0 {
+        Some(value.vce_index)
+    } else {
+        None
+    };
+
+    (
+        Ident(value.vce_function, idx),
+        Entry {
+            eax: value.vce_eax,
+            ebx: value.vce_ebx,
+            ecx: value.vce_ecx,
+            edx: value.vce_edx,
+        },
+    )
+}
+
+pub struct Iter<'a>(std::collections::btree_map::Iter<'a, Ident, Entry>);
+impl<'a> Iterator for Iter<'a> {
+    type Item = (&'a Ident, &'a Entry);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next()
+    }
+}
+
+/// Specialize a set of cpuid leafs for provided attributes.
+///
+/// This includes things such as a CPU topology (cores/threads/etc), a given
+/// vCPU ID (APIC, core/thread ID, etc), or other info tidbits.
+#[derive(Default)]
+pub struct Specializer {
+    has_smt: bool,
+    num_vcpu: Option<NonZeroU8>,
+    vcpuid: Option<i32>,
```

Comment:
Do we ever expect these to be None? I'd almost be inclined to make `num_vcpu` a required parameter to `new` and `vcpuid` a required parameter to `execute`.

Reply:
I'm still curious to see how the interface (and usage) gets fleshed out, so I'm going to leave it for now.  We can circle back to making it more strict in the future.